File: C++/50.Pow(x,n).cpp
class Solution {
public:
    double myPow(double x, int n) {
        //check the sign of n
        int sign=0;
        
        if(x!=1.0 && x!=-1.0){
            if(n>0){
                sign=1;
           }else if(n<0){
                sign=-1;
           }

           double con=1;

 
           if(sign==1 and x!=1 ){
               while(n!=1){
                    if(n%2==1){
                      con*=x;
              
                      n-=1;
                    }
                    n=n/2;
                    x=x*x;
                   cout<<x;
                   
                }
               x=x*con;
            
         
            
          }else if(sign==-1 and x!=1){
               x=(1/x);
               cout<<x;
               while(n!=-1){
                    if(n%2==-1){
                      con*=x;
                      n+=1;
                    }
                    n=n/2;
                    x=x*x;
                   
                }
               x=x*con;
               if(x<0){
                   x=x*-1;
               }
               
            }else if(sign==0 and x!=1){
                   x=1;
               }
            
        }else if(x==1){
                x=1;
        }else if(x==-1){

            if(n&1){
                x=-1;
            }else{
                x=1;
            }
        }
        return x;
    
}
    };

File: python/50. Pow(x, n).py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        
        if(n == 0):
            return 1
        if(n < 0):
            x = (1/x)
            n *= -1

        if(x == 1):
            return 1
        elif(x == -1):
            if(n%2 ==1):
                return -1
            else:
                return 1
        
        org = x
        times = 1
        while(times < n):
  
            if(times*2 <= n):
                times*=2
                x = x*x
            else:
                if(org < 1 and x== 0):
                    break

                times+=1
                x *= org
        

        return x
